[Frontend] Add section for using JSX with Vue.js (#11346)
symfony/symfony-docs: frontend/encore/vuejs.rst

See :doc:`/frontend/encore/dev-server` for more details.

JSX Support
-----------

> You can enable `JSX with Vue.js`_ by configuring the 2nd parameter of ``.enableVueLoader`` method:

Reviewer: ```suggestion
You can enable `JSX with Vue.js`_ by configuring the 2nd parameter of the ``.enableVueLoader`` method:
```

.. code-block:: diff

    // webpack.config.js
    // ...

    Encore
        // ...
        .addEntry('main', './assets/main.js')

    -     .enableVueLoader()
    +     .enableVueLoader(() => {}, {
    +         useJsx: true
    +     })
    ;

> Then restart Encore. When you do, it will give you a command you can run to

Reviewer: ```suggestion
Afterwards rebuild your frontend assets with Encore. When you do it will show an error message helping you
```

install any missing dependencies. After running that command and restarting
Encore, you're done!

Your ``.jsx`` files will now be transformed through ``@vue/babel-preset-jsx``.

Using styles
~~~~~~~~~~~~

You can't use ``<style>`` in ``.jsx`` files.

As a workaround, you can import ``.css``, ``.scss``, ``.less`` and ``.styl`` files manually:

.. code-block:: js

    // App.jsx

    import './App.css'

    export default {
        name: 'App',
        render() {
            return (
                <div>
                    ...
                </div>
            )
        }
    }

> Note that importing styles like this make them global.

Reviewer: Use a note block for this
```suggestion
.. note::
```

See :doc:`/frontend/encore/dev-server` for more details.

JSX Support
-----------

> You can enable `JSX with Vue.js`_ by configuring the 2nd parameter of the ``.enableVueLoader`` method:

Reviewer 2: ```suggestion
You can enable `JSX with Vue.js`_ by configuring the 2nd parameter of the ``.enableVueLoader()`` method:
```

.. code-block:: diff

    // webpack.config.js
    // ...

    Encore
        // ...
        .addEntry('main', './assets/main.js')

    -     .enableVueLoader()
    +     .enableVueLoader(() => {}, {
    +         useJsx: true
    +     })
    ;

> Afterwards rebuild your frontend assets with Encore. When you do it will show an error message helping you

Reviewer 2: ```suggestion
Next, run or restart Encore. When you do, you will see an error message helping you
```

install any missing dependencies. After running that command and restarting
Encore, you're done!

Your ``.jsx`` files will now be transformed through ``@vue/babel-preset-jsx``.

Using styles
~~~~~~~~~~~~

You can't use ``<style>`` in ``.jsx`` files.

> As a workaround, you can import ``.css``, ``.scss``, ``.less`` and ``.styl`` files manually:

Reviewer 2: Put this sentence on the previous paragraph, instead of its own.

Reviewer 2: ```suggestion
As a workaround, you can import ``.css``, ``.scss``, etc, files manually:
```

> .. code-block:: js

Reviewer 3: Not sure it is “js” or “JavaScript”

Author: Both work

Reviewer 3: Sorry for not being clear, I was on a phone. This was more or less a question to the team.
We prefer „javascript“ here, thank you 😊

Author: Oh okay, I will change it ASAP :)

Reviewer 3: Great, thank you very much

    // App.jsx

    import './App.css'

    export default {
        name: 'App',
        render() {
            return (
                <div>
                    ...
                </div>
            )
        }
    }

.. note::

    Importing styles this way make them global.
    See the next section for scoping them to your component.

Using Scoped Styles
~~~~~~~~~~~~~~~~~~~

> Same, you can't use `Scoped Styles`_ (``<style scoped>``) in ``.jsx`` files.

Reviewer 2: ```suggestion
You also can't use `Scoped Styles`_ (``<style scoped>``) in ``.jsx`` files.
```

> As a workaround, you can use `CSS Modules`_ by suffixing import paths with ``?module``:

Reviewer 2: Move to previous paragraph

.. code-block:: js

    // Component.jsx

    import styles from './Component.css?module' // suffix with "?module"

    export default {
        name: 'Component',
        render() {
            return (
                <div>
                    <h1 class={styles.title}>
                        Hello World
                    </h1>
                </div>
            )
        }
    }

.. code-block:: css

    /* Component.css */

    .title {
        color: red
    }

The output will be something like ``<h1 class="h1_a3dKp">Hello World</h1>``.

Using images
~~~~~~~~~~~~

You can't use ``<img src="./image.png">`` in ``.jsx`` files.

> As a workaround, you can import them with ``require()`` function:

Reviewer 2: Previous paragraph
